tests/subsys/bluetooth/ble_bas/src/unity_test.c

```diff
@@ -0,0 +1,407 @@
+{
+	TEST_ASSERT_EQUAL(BLE_BAS_EVT_NOTIFICATION_DISABLED, evt->evt_type);
+	evt_handler_called = true;
+}
+
+void test_ble_bas_on_ble_evt_test(void)
+{
+	ble_evt_t evt = {0};
+
+	ble_bas.can_notify = true;
+	ble_bas.battery_level_handles.cccd_handle = 0x1234;
+	evt.evt.gatts_evt.params.write.handle = 0x1234;
+	evt.evt.gatts_evt.params.write.len = 2;
+
+	evt.header.evt_id = BLE_GATTS_EVT_WRITE;
+	evt.evt.gatts_evt.params.write.data[0] = BLE_GATT_HVX_NOTIFICATION;
+	ble_bas.evt_handler = ble_bas_evt_handler_notif_enabled;
+	ble_bas_on_ble_evt(&evt, &ble_bas);
+	TEST_ASSERT_TRUE(evt_handler_called);
+
+	evt_handler_called = false;
+	evt.evt.gatts_evt.params.write.data[0] = BLE_GATT_HVX_INDICATION;
+	ble_bas.evt_handler = ble_bas_evt_handler_notif_disable;
+	ble_bas_on_ble_evt(&evt, &ble_bas);
+	TEST_ASSERT_TRUE(evt_handler_called);
+
+	evt_handler_called = false;
+	ble_bas.can_notify = false;
+	ble_bas.evt_handler = ble_bas_evt_handler;
+	ble_bas_on_ble_evt(&evt, &ble_bas);
+	TEST_ASSERT_FALSE(evt_handler_called);
+
+	ble_bas.can_notify = true;
+	ble_bas.battery_level_handles.cccd_handle = 0x5678;
+	ble_bas_on_ble_evt(&evt, &ble_bas);
+	TEST_ASSERT_FALSE(evt_handler_called);
+
+	ble_bas.battery_level_handles.cccd_handle = 0x1234;
+	evt.evt.gatts_evt.params.write.len = 1;
+	ble_bas_on_ble_evt(&evt, &ble_bas);
+	TEST_ASSERT_FALSE(evt_handler_called);
+}
+
+void test_ble_bas_init_efault(void)
+{
+	int ret;
+	struct ble_bas_config bas_config = {
+		.evt_handler = ble_bas_evt_handler,
+	};
+
+	ret = ble_bas_init(NULL, &bas_config);
+	TEST_ASSERT_EQUAL(-EFAULT, ret);
+	TEST_ASSERT_NOT_EQUAL(bas_config.evt_handler, ble_bas.evt_handler);
+
+	ret = ble_bas_init(&ble_bas, NULL);
+	TEST_ASSERT_EQUAL(-EFAULT, ret);
+	TEST_ASSERT_NOT_EQUAL(bas_config.evt_handler, ble_bas.evt_handler);
+}
+
+void test_ble_bas_init_einval(void)
+{
+	int ret;
+	struct ble_bas_config bas_config = {
+		.evt_handler = ble_bas_evt_handler,
+	};
+
+	static struct {
+		uint8_t report_id;
+		uint8_t report_type;
+	} report_ref = {.report_id = 1, .report_type = 0x01};
+
+	bas_config.report_ref = (void *)&report_ref;
+
+	__cmock_sd_ble_gatts_service_add_ExpectAnyArgsAndReturn(NRF_ERROR_INVALID_PARAM);
+	ret = ble_bas_init(&ble_bas, &bas_config);
+	TEST_ASSERT_EQUAL(-EINVAL, ret);
+
+	__cmock_sd_ble_gatts_service_add_ExpectAnyArgsAndReturn(NRF_SUCCESS);
+	__cmock_sd_ble_gatts_characteristic_add_ExpectAnyArgsAndReturn(NRF_ERROR_INVALID_PARAM);
+	ret = ble_bas_init(&ble_bas, &bas_config);
+	TEST_ASSERT_EQUAL(-EINVAL, ret);
+
+	__cmock_sd_ble_gatts_service_add_ExpectAnyArgsAndReturn(NRF_SUCCESS);
+	__cmock_sd_ble_gatts_characteristic_add_ExpectAnyArgsAndReturn(NRF_SUCCESS);
+	__cmock_sd_ble_gatts_descriptor_add_ExpectAnyArgsAndReturn(NRF_ERROR_INVALID_PARAM);
+	ret = ble_bas_init(&ble_bas, &bas_config);
+	TEST_ASSERT_EQUAL(-EINVAL, ret);
+}
+
+static uint32_t sd_ble_gatts_characteristic_add_stub(uint16_t service_handle,
```

Comment:
Place stubs at top of file and use stub as prefix
```suggestion
static uint32_t stub_sd_ble_gatts_characteristic_add(uint16_t service_handle,
```